libecho/src/BitAudioConverter.h

```diff
@@ -0,0 +1,39 @@
+#ifndef ECHOCONNECT_BIT_AUDIO_CONVERTER_H
+#define ECHOCONNECT_BIT_AUDIO_CONVERTER_H
+
+#include "IAudioConverter.h"
+
+#include <cmath>
+
+template <typename T>
+class BitAudioConverter : public IAudioConverter<T> {
+public:
+    BitAudioConverter(QAudioFormat inputFormat, QAudioFormat outputFormat, double windowLength, int lo, int hi)
+        : IAudioConverter<T>(inputFormat, outputFormat, windowLength), loFreq(lo), hiFreq(hi) {}
+
+private:
+    const int loFreq, hiFreq;
+
+    std::vector<T> encode_byte(uint8_t data) override {
+        std::vector<T> result;
+        for (int i = 0; i < 8; i++) {
+            int bit = (data >> i) & 1;
+            int freq = (bit == 1) ? hiFreq : loFreq;
+            for (int t = 0; t < this->outputFormat.sampleRate() * this->windowLength; t++) {
+                result.emplace_back(audio_wave(freq, t));
+            }
+        }
+        return result;
+    }
+
+    uint8_t decode_byte(std::vector<char>) override {
+        return 0;
+    }
+
+    T audio_wave(int frequency, int time) {
+        return sin((double)frequency * time / this->outputFormat.sampleRate()) *
```

Comment:
czym jest zwracany typ T? jeżeli zakładamy, że to jeden z podstawowych typów zmiennoprzecinkowych (float, double, long double), to raczej powinniśmy frequency castować konkretnie na T, a nie na double, aby uniknąć wielokrotnego castowania

Reply:
Typ zwracany to char/short. A ten cast na doubla to chyba po to żeby do sin przekazac doubla, nie ma nic wspolnego z typem zwracanym

Reply:
W zależności od tego, jak będzie ustawione SampleSize oraz SampleType w AudioFormatFactory będą to różne typy, nie zawsze zmiennoprzecinkowe (na ten moment jest to int16_t). W związku z tym nie możemy castować po prostu na T.